test/src/test_async.rs

```diff
@@ -0,0 +1,99 @@
+use std::cell::RefCell;
+
+use gdnative::asn::{Async, AsyncMethod, Spawner};
+use gdnative::prelude::*;
+
+pub(crate) fn run_tests() -> bool {
+    // Relevant tests in GDScript
+    true
+}
+
+thread_local! {
+    static EXECUTOR: &'static SharedLocalPool = {
+        Box::leak(Box::new(SharedLocalPool::default()))
+    };
+}
+
+pub(crate) fn register(handle: InitHandle) {
+    gdnative::asn::register_runtime(&handle).unwrap();
+    gdnative::asn::set_executor(EXECUTOR.with(|e| *e)).unwrap();
+
+    handle.add_class::<AsyncMethods>();
+    handle.add_class::<AsyncExecutorDriver>();
+}
+
+#[derive(Default)]
+struct SharedLocalPool {
+    pool: RefCell<futures::executor::LocalPool>,
+}
+
+impl futures::task::LocalSpawn for SharedLocalPool {
+    fn spawn_local_obj(
+        &self,
+        future: futures::task::LocalFutureObj<'static, ()>,
+    ) -> Result<(), futures::task::SpawnError> {
+        self.pool.borrow_mut().spawner().spawn_local_obj(future)
+    }
+}
+
+#[derive(NativeClass)]
+#[inherit(Node)]
+struct AsyncExecutorDriver;
+
+impl AsyncExecutorDriver {
+    fn new(_owner: &Node) -> Self {
+        AsyncExecutorDriver
+    }
+}
+
+#[methods]
+impl AsyncExecutorDriver {
+    #[export]
+    fn _process(&self, _owner: &Node, _delta: f64) {
+        EXECUTOR.with(|e| e.pool.borrow_mut().run_until_stalled());
+    }
+}
+
+#[derive(NativeClass)]
+#[inherit(Reference)]
+#[register_with(register_methods)]
+struct AsyncMethods;
+
+#[methods]
+impl AsyncMethods {
+    fn new(_owner: TRef<Reference>) -> Self {
+        AsyncMethods
+    }
+}
+
+struct ResumeAddFn;
+
+impl AsyncMethod<AsyncMethods> for ResumeAddFn {
+    fn spawn_with(&self, spawner: Spawner<'_, AsyncMethods>) {
+        spawner.spawn(|ctx, _this, mut args| {
+            let a = args.read::<i32>().get().unwrap();
+            let obj = args.read::<Ref<Object>>().get().unwrap();
+            let name = args.read::<GodotString>().get().unwrap();
```

Comment:
If this was not test code, how would we expect someone to error out if the args are the wrong type?

Reply:
Use a `FromVarargs` structure, log the error from `read_many` and return `Nil`. See `gdnative_core::nativescript::init::method::StaticArgsMethod`.

This `AsyncMethod` trait here is a lower-level interface comparable to `gdnative_core::nativescript::init::method::Method`. End-users aren't expected to use this directly unless they need to do some really complicated run-time magic, at which point we should get out of the way and leave proper error handling to them. An interface comparable to `StaticArgsMethod` will be added later as we figure out how we want to support this in `#[methods]`.